### DISCUSSION TITLE
Workflow fails when running steps in a container

### DISCUSSION THREAD

**Discussion by curthasselschwert** [community/community#25094] [Actions]
<p>I’ve just started setting up a workflow for one of my Elixir projects. I began with a simple workflow that just echos text in the shell. Here is that worklow:</p>
<pre><code>name: CI

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v1
    - name: Run a one-line script
      run: echo Hello, world!
    - name: Run a multi-line script
      run: |
        echo Add other actions to build,
        echo test, and deploy your project.
</code></pre>
<p>This workflow completes successfully.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/2e3402c23db6f7f011d1b70541d8e3c52d3754a3.png" data-download-href="/uploads/short-url/6AJoz39ZgezHEdUeYwpxyDrtWo3.png?dl=1" title="Screen Shot 2019-08-21 at 3.24.13 PM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/82232/181088731-5e63ad71-d34c-47cd-b56b-40da0510e6e7.png" alt="Screen Shot 2019-08-21 at 3.24.13 PM.png" data-base62-sha1="6AJoz39ZgezHEdUeYwpxyDrtWo3" width="690" height="248" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/2e3402c23db6f7f011d1b70541d8e3c52d3754a3_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-08-21 at 3.24.13 PM.png</span><span class="informations">816×294 5.91 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Next, I attempted to run the same commands inside a container with this workflow definition:</p>
<pre><code>name: CI

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    container: elixir:1.6.2-alpine

    steps:
    - uses: actions/checkout@v1
    - name: Run a one-line script
      run: echo Hello, world!
    - name: Run a multi-line script
      run: |
        echo Add other actions to build,
        echo test, and deploy your project.
</code></pre>
<p>This workflow fails with the error <code>Error response from daemon: Container XXXXXX is not running</code>.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/7c969ed19bb211ea4f62a07a4b22dc91f409f8f8.png" data-download-href="/uploads/short-url/hM9UmhTJTpZyWIuKEeDkIDV1X5K.png?dl=1" title="Screen Shot 2019-08-21 at 3.26.10 PM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/82232/181088732-05b1dbef-1d3b-49a2-a2c0-d38c1fa77dd9.png" alt="Screen Shot 2019-08-21 at 3.26.10 PM.png" data-base62-sha1="hM9UmhTJTpZyWIuKEeDkIDV1X5K" width="690" height="383" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/7c969ed19bb211ea4f62a07a4b22dc91f409f8f8_2_690x383.png, https://aws1.discourse-cdn.com/github/original/2X/7/7c969ed19bb211ea4f62a07a4b22dc91f409f8f8.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/7c969ed19bb211ea4f62a07a4b22dc91f409f8f8.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/7c969ed19bb211ea4f62a07a4b22dc91f409f8f8_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-08-21 at 3.26.10 PM.png</span><span class="informations">832×463 26.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I’ve just started to work with GitHub Actions, so maybe there is something I’ve overlooked. Can anyone help me resolve this issue? According to <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idcontainer" rel="nofollow noopener">Workflow syntax for GitHub Actions</a>, in the <em>Container</em> section: “A container to run any steps in a job that don’t already specify a container.” My assumption is that the container would be started and the steps would be run inside the Elixir Alpine Docker image.</p>

**Comment by willfore** [comment]
<p>I’m having the exact same issue currently. Did you find a work around?</p>

**Comment by curthasselschwert** [comment]
<p>I haven’t figured out a solution yet. GitHub must be having issues because now the workflow isn’t even running.</p>

**Comment by Aleksion** [comment]
<p>I’m having the eact same one as well. </p>

**Comment by juulSme** [comment]
<p>I have the same issue. It works when you use “-slim” images. Somehow, the alpine container is not detected as running.</p>

**Comment by juulSme** [comment, marked as answer]
<pre><code>ci_1_9:
    name: Elixir 1.9 CI
    runs-on: ubuntu-latest

    container:
      # alpine causes errors in the steps, the container exits immediately, apparently
      image: elixir:1.9-slim
      env:
        MIX_ENV: test

    steps:
      - name: Checkout code
        uses: actions/checkout@master
      - name: Display build environment
        run: printenv
      - name: Install dependencies
        run: mix do local.hex --force, local.rebar --force, deps.get
      - name: Compile code (warnings as errors)
        run: mix compile --warnings-as-errors
      - name: Check code formatting
        run: mix format --check-formatted
      - name: Run tests (with coverage)
        run: mix test --cover
      - name: Create docs
        run: mix docs

  ci_1_9_alpine:
    name: Elixir 1.9 CI Alpine
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@master
      - name: Display build environment
        run: printenv
      - name: Install dependencies, compile, check formatting, test and create docs
        uses: docker://elixir:1.9-alpine
        env:
          MIX_ENV: test
        with:
          entrypoint: /bin/sh
          args: |
            -c "\
            mix do local.hex --force, local.rebar --force, deps.get &amp;&amp; \
            mix compile --warnings-as-errors &amp;&amp; \
            mix format --check-formatted &amp;&amp; \
            mix test --cover &amp;&amp; \
            mix docs"
</code></pre>
<p>Either of these work. I prefer the “slim” one, because in the alpine one (in this form) all commands get thrown together in one github actions step.</p>

**Comment by laszlocph** [comment]
<p>This solved it for me too. Seems Alpine images has a bug in the runner.</p>

**Comment by PhilippHeuer** [comment]
<p>I had the same issue, i tried to look into what is happening a bit and it mounts tools from the host into the container and then overwrites the entrypoint with node to run setInterval to keep the container running.</p>
<pre><code>--entrypoint "/__e/node12/bin/node" yourdockerimage -e "setInterval(function(){}, 24 * 60 * 60 * 1000);"
</code></pre>
<p>This is naturally a bad idea, as those binaries aren’t compiled for alpine.</p>
<p>I also found some requirements on the Azure DevOps Site, which are prob. in effect for GitHub Actions too - but i didnt find a notice about this in the gh action docs:</p>
<pre><code>Bash
glibc-based
Can run Node.js (which the agent provides)
</code></pre>
<p>Seems pretty overkill though, to have dependencies on glibc and nodejs just to run setInterval to keep the container running, there are better ways like for example: _/bin/sh -c “tail -f /dev/null” or just by using sleep _that do not have any dependencies and should also be usable everywhere where bash is available.</p>
<p>Sadly we can’t affect the entrypoint / args passed to the main container, so the only possible way to workaround this is to either install glibc and all dependencies into the alpine image (not viable), or to not use container and define the image to use for every single command (bloats workflow up quite a bit).</p>
<p>Im working on a different workaround that i will share later.</p>

**Comment by juulSme** [comment]
<p>Kudos for the research :slight_smile: it does seem weird that they don’t just tail to dev/null to keep the container up, that’s a pretty standard way of operating.</p>

**Comment by teohhanhui** [comment]
<p>This is most definitely not “solved”. Using the <code>alpine</code> variant of the official Docker images (in my case <code>php</code>) seems to be causing a failure.</p>

**Comment by dakale** [comment]
<p>I am actually working on making this exact change. In earlier versions, we were running actions such as git checkout by default, and they were written in JavaScript, so naturally we had a higher level of dependency on Node here. After shedding that, we now have legitimate scenarios like the ones described here that shouldn’t need Node at all. Expect that update soon!</p>

**Comment by chrispat** [comment]
<p>This should now be fully supported</p>
<p><a href="https://github.blog/changelog/2019-09-18-improvements-to-github-actions/" rel="nofollow noopener">https://github.blog/changelog/2019-09-18-improvements-to-github-actions/</a> </p>

**Comment by sumanstats** [comment]
<p>Is it possible to run same commands in three docker containers with elixir 1.9, 1.8 and 1.7 using matrix strategy?</p>
